Content.Server/Flight/FlightSystem.cs

```diff
@@ -0,0 +1,151 @@
+            SubscribeLocalEvent<FlightComponent, EntityZombifiedEvent>(OnZombified);
+            SubscribeLocalEvent<FlightComponent, KnockedDownEvent>(OnKnockedDown);
+            SubscribeLocalEvent<FlightComponent, StunnedEvent>(OnStunned);
+            SubscribeLocalEvent<FlightComponent, SleepStateChangedEvent>(OnSleep);
+        }
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            var query = EntityQueryEnumerator<FlightComponent>();
+            while (query.MoveNext(out var uid, out var component))
+            {
+                if (!component.On)
+                    continue;
+
+                component.TimeUntilFlap -= frameTime;
+
+                if (component.TimeUntilFlap <= 0f)
+                {
+                    _audio.PlayPvs(component.FlapSound, uid);
+                    component.TimeUntilFlap = component.FlapInterval;
+                }
+            }
+        }
+
+        #region Core Functions
+        private void OnToggleFlight(EntityUid uid, FlightComponent component, ToggleFlightEvent args)
+        {
+            // If the user isnt flying, we check for conditionals and initiate a doafter.
+            if (!component.On)
+            {
+                if (CanFly(uid, component))
+                {
+                    var doAfterArgs = new DoAfterArgs(EntityManager, uid, component.ActivationDelay, new FlightDoAfterEvent(), uid, target: uid)
+                    {
+                        BlockDuplicate = true,
+                        BreakOnTargetMove = true,
+                        BreakOnUserMove = true,
+                        BreakOnDamage = true,
+                        NeedHand = true
+                    };
+
+                    if (!_doAfter.TryStartDoAfter(doAfterArgs))
+                    {
+                        return;
+                    }
+                }
+            }
+            else
+            {
+                ToggleActive(uid, false, component);
+            }
+        }
+
+        private void OnFlightDoAfter(EntityUid uid, FlightComponent component, FlightDoAfterEvent args)
+        {
+            if (args.Handled || args.Cancelled)
+                return;
+
+            ToggleActive(uid, true, component);
+            args.Handled = true;
+        }
+
+        #endregion
+
+        #region Conditionals
+
+        private bool CanFly(EntityUid uid, FlightComponent component)
+        {
+            var cuffed = TryComp<CuffableComponent>(uid, out var cuffableComp) && !cuffableComp.CanStillInteract;
+            var zombified = TryComp<ZombieComponent>(uid, out var _);
+
+            // Tell the user that they can not fly.
+            if (cuffed || zombified)
+            {
+                _popupSystem.PopupEntity(Loc.GetString("no-flight-while-restrained"), uid, uid, PopupType.Medium);
+                return false;
+            }
+
+            return true;
+        }
```

Comment:
```suggestion
        private bool CanFly(EntityUid uid, FlightComponent component)
        {
            if (TryComp<CuffableComponent>(uid, out var cuffableComp) && !cuffableComp.CanStillInteract)
            {
                _popupSystem.PopupEntity(Loc.GetString("no-flight-while-restrained"), uid, uid, PopupType.Medium);
                return false;
            }
            
            if (HasComp<ZombieComponent>(uid))
            {
                _popupSystem.PopupEntity(Loc.GetString("no-flight-while-zombified"), uid, uid, PopupType.Medium);
                return false;
            }
            return true;
        }
```
You're also going to need to add a string for no-flight-while-zombified.